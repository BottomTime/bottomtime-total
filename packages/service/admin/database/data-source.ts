import path from 'path';
import { DataSource } from 'typeorm';

export async function getDataSource(
  url: string,
  requireSsl: boolean,
): Promise<DataSource> {
  const client = new DataSource({
    type: 'postgres',
    url,
    entities: [path.resolve(__dirname, '../../src/data/**/*.entity.ts')],
    ssl: requireSsl,
  });
  return await client.initialize();
}
