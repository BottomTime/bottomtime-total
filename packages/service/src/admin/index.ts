export * from './admin.service';

export * from './admin-tanks.controller';
export * from './admin-users.controller';

export * from './admin.module';
