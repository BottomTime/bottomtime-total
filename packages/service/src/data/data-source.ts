import path from 'path';
import { DataSource, DataSourceOptions } from 'typeorm';

import { Config } from '../config';

export const PostgresDataSourceOptions: DataSourceOptions = {
  type: 'postgres',
  url: Config.postgresUri,
  entities: [path.join(__dirname, './**/*.entity.ts')],
  synchronize: false,
};

export async function initDataSource(): Promise<DataSource> {
  const AppDataSource = new DataSource(PostgresDataSourceOptions);
  return await AppDataSource.initialize();
}
