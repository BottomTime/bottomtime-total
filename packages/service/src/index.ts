import { Config } from './config';
import { createApp } from './create-app';
import { createLogger } from './logger';

const log = createLogger(Config.logLevel);

createApp(log)
  .then(async (app) => {
    await app.listen(Config.port);

    log.info(
      `ðŸŽ‰ Service has successfully started and is listening on port ${Config.port}. ðŸŽ‰`,
    );
  })
  .catch((error) => {
    log.fatal('[SERVICE] Failed to start service... Shutting down...', error);
    process.exit(1);
  });
