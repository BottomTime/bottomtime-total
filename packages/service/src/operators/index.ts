export * from './operator';
export * from './operator-review';

export * from './operators.module';
