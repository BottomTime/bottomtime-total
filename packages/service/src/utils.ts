import {
  DepthDTO,
  DepthUnit,
  PressureDTO,
  PressureUnit,
  TemperatureDTO,
  TemperatureUnit,
  WeightDTO,
  WeightUnit,
} from '@bottomtime/api';

export class UnitUtils {
  static convertDepth(depth: DepthDTO, expectedUnit: DepthUnit): DepthDTO {
    if (expectedUnit === depth.unit) return depth;

    if (expectedUnit === DepthUnit.Meters) {
      return {
        depth: depth.depth * 0.3048,
        unit: DepthUnit.Meters,
      };
    } else {
      return {
        depth: depth.depth / 0.3048,
        unit: DepthUnit.Feet,
      };
    }
  }

  static convertTemperature(
    temperature: TemperatureDTO,
    expectedUnit: TemperatureUnit,
  ): TemperatureDTO {
    if (expectedUnit === temperature.unit) return temperature;

    if (expectedUnit === TemperatureUnit.Celsius) {
      return {
        temperature: (temperature.temperature - 32) * (5 / 9),
        unit: TemperatureUnit.Celsius,
      };
    } else {
      return {
        temperature: temperature.temperature * (9 / 5) + 32,
        unit: TemperatureUnit.Fahrenheit,
      };
    }
  }

  static convertPressure(
    pressure: PressureDTO,
    expectedUnit: PressureUnit,
  ): PressureDTO {
    if (expectedUnit === pressure.unit) return pressure;

    if (expectedUnit === PressureUnit.Bar) {
      return {
        pressure: pressure.pressure * 0.0689475729,
        unit: PressureUnit.Bar,
      };
    } else {
      return {
        pressure: pressure.pressure / 0.0689475729,
        unit: PressureUnit.PSI,
      };
    }
  }

  static convertWeight(weight: WeightDTO, expectedUnit: WeightUnit): WeightDTO {
    if (expectedUnit === weight.unit) return weight;

    if (expectedUnit === WeightUnit.Kilograms) {
      return {
        weight: weight.weight * 2.20462,
        unit: WeightUnit.Kilograms,
      };
    } else {
      return {
        weight: weight.weight / 2.20462,
        unit: WeightUnit.Pounds,
      };
    }
  }
}
