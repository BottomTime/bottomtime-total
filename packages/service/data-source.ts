import path from 'path';
import { DataSource } from 'typeorm';

import { PostgresDataSourceOptions } from './src/data/data-source';

export default new DataSource({
  ...PostgresDataSourceOptions,
  migrations: [path.join(__dirname, './migrations/*.ts')],
});
