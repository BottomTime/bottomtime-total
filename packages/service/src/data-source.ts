import 'reflect-metadata';
import { DataSourceOptions } from 'typeorm';

import { Config } from './config';
import * as Entities from './data';

export const PostgresDataSourceOptions: DataSourceOptions = {
  type: 'postgres',
  url: Config.postgresUri,
  entities: Object.values(Entities),
  synchronize: false,
};
