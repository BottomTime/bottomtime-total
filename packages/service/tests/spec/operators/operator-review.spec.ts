import { AccountTier, LogBookSharing } from '@bottomtime/api';

import { Repository } from 'typeorm';

import {
  OperatorEntity,
  OperatorReviewEntity,
  UserEntity,
} from '../../../src/data';
import { OperatorReview } from '../../../src/operators';
import { dataSource } from '../../data-source';
import { createTestOperator, createTestUser } from '../../utils';

const TestOperatorOwner = createTestUser();
const TestOperator = createTestOperator(TestOperatorOwner);
const TestCreator: UserEntity = createTestUser({
  id: '3978ee62-9743-4919-af33-254efa821654',
  accountTier: AccountTier.Basic,
  memberSince: new Date('2025-01-01T00:00:00Z'),
  username: 'TestyMcTestFace',
  logBookSharing: LogBookSharing.Public,
  avatar: 'https://example.com/avatar.jpg',
  name: 'Testy McTestFace',
  location: 'Testville, Testland',
});
const TestReviewData: OperatorReviewEntity = {
  id: '8f5106e0-e9be-4c37-9fd0-aadf6f964ddb',
  comments:
    'This operator was great! Great staff, great service, great experience! They showed me the secret handshake and everything!',
  createdAt: new Date('2025-01-02T10:42:06-05:00'),
  creator: TestCreator,
  rating: 4.7,
  operator: TestOperator,
  updatedAt: new Date('2025-01-02T10:42:06-05:00'),
  logEntry: null,
};

describe('OperatorReview class', () => {
  let Operators: Repository<OperatorEntity>;
  let Users: Repository<UserEntity>;
  let Reviews: Repository<OperatorReviewEntity>;

  let reviewData: OperatorReviewEntity;
  let review: OperatorReview;

  beforeAll(() => {
    Operators = dataSource.getRepository(OperatorEntity);
    Users = dataSource.getRepository(UserEntity);
    Reviews = dataSource.getRepository(OperatorReviewEntity);
  });

  beforeEach(async () => {
    await Users.save([TestCreator, TestOperatorOwner]);
    await Operators.save(TestOperator);

    reviewData = { ...TestReviewData };
    review = new OperatorReview(Reviews, reviewData);
  });

  it('should return properties correctly', () => {
    expect(review.id).toBe(reviewData.id);
    expect(review.comments).toBe(reviewData.comments);
    expect(review.createdAt).toEqual(reviewData.createdAt);
    expect(review.updatedAt).toEqual(reviewData.updatedAt);
    expect(review.rating).toBe(reviewData.rating);
    expect(review.creator).toMatchSnapshot();
  });

  it('will return null properties as undefined', () => {
    reviewData.comments = null;
    expect(review.comments).toBeUndefined();
  });

  it('will allow properties to be updated', () => {
    review.comments = 'Updated comments';
    review.rating = 3.2;

    expect(review.comments).toBe('Updated comments');
    expect(review.rating).toBe(3.2);
  });

  it('will render as a JSON DTO', () => {
    expect(review.toJSON()).toMatchSnapshot();
  });

  it('will fail silently when deleting a review that does not exist', async () => {
    await expect(review.delete()).resolves.toBe(false);
  });

  it('will delete a review', async () => {
    await Reviews.save(reviewData);
    await expect(review.delete()).resolves.toBe(true);
    await expect(Reviews.findOneBy({ id: reviewData.id })).resolves.toBeNull();
  });

  it('will save a new review', async () => {
    await review.save();
    const saved = await Reviews.findOneOrFail({
      where: { id: reviewData.id },
      relations: ['creator', 'operator'],
    });
    expect(saved.comments).toBe(reviewData.comments);
    expect(saved.createdAt).toEqual(reviewData.createdAt);
    expect(saved.creator.id).toBe(TestCreator.id);
    expect(saved.operator.id).toBe(TestOperator.id);
    expect(saved.rating).toBe(reviewData.rating);
    expect(saved.updatedAt).toEqual(reviewData.updatedAt);
  });

  it('will update an existing review', async () => {
    await Reviews.save(reviewData);
    review.comments = 'Updated comments';
    review.rating = 3.2;
    await review.save();

    const saved = await Reviews.findOneOrFail({
      where: { id: reviewData.id },
    });
    expect(saved.comments).toBe('Updated comments');
    expect(saved.rating).toBe(3.2);
  });
});
