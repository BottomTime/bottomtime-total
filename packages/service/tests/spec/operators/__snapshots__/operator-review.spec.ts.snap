// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OperatorReview class should return properties correctly 1`] = `
{
  "accountTier": 0,
  "avatar": "https://example.com/avatar.jpg",
  "location": "Testville, Testland",
  "logBookSharing": "public",
  "memberSince": 1735689600000,
  "name": "Testy McTestFace",
  "userId": "3978ee62-9743-4919-af33-254efa821654",
  "username": "TestyMcTestFace",
}
`;

exports[`OperatorReview class will render as a JSON DTO 1`] = `
{
  "comments": "This operator was great! Great staff, great service, great experience! They showed me the secret handshake and everything!",
  "createdAt": 1735832526000,
  "creator": {
    "accountTier": 0,
    "avatar": "https://example.com/avatar.jpg",
    "location": "Testville, Testland",
    "logBookSharing": "public",
    "memberSince": 1735689600000,
    "name": "Testy McTestFace",
    "userId": "3978ee62-9743-4919-af33-254efa821654",
    "username": "TestyMcTestFace",
  },
  "id": "8f5106e0-e9be-4c37-9fd0-aadf6f964ddb",
  "rating": 4.7,
  "updatedAt": 1735832526000,
}
`;
